accepted/2024/support-nuget-authentication-plugins-dotnet-tools.md

```diff
@@ -0,0 +1,210 @@
+
+Another motivating factor for this work is the current story for installing these credential providers, which is not ideal. For instance, let's consider the two cross-platform authentication plugins that I am aware of while writing this specification:
+
+1. **Azure Artifacts Credential Provider** - The [setup](https://github.com/microsoft/artifacts-credprovider/tree/master?tab=readme-ov-file#setup) instructions vary based on the platform.
+2. **AWS CodeArtifact Credential Provider** - The AWS team has developed a [.NET tool](https://docs.aws.amazon.com/codeartifact/latest/ug/nuget-cli.html#nuget-configure-cli) to facilitate authentication with their private NuGet feeds. In their current implementation, they've added a subcommand, `codeartifact-creds install`, which copies the credential provider to the NuGet plugins folder. The log below shows all possible subcommands. However, plugin authors could leverage the .NET SDK to allow their customers to install, uninstall, or update the NuGet plugins more efficiently.
+
+```log
+~\.nuget\plugins\tools
+❯ .\dotnet-codeartifact-creds.exe
+Required command was not provided.
+
+Usage:
+  dotnet-codeartifact-creds [options] [command]
+
+Options:
+  --version         Show version information
+  -?, -h, --help    Show help and usage information
+
+Commands:
+  install      Installs the AWS CodeArtifact NuGet credential provider into the NuGet plugins folder.
+  configure    Sets or Unsets a configuration
+  uninstall    Uninstalls the AWS CodeArtifact NuGet credential provider from the NuGet plugins folder.
+```
+
+## Explanation
+
+### Functional explanation
+
+A deployment solution for .NET is the [.NET tools](https://learn.microsoft.com/dotnet/core/tools). These tools provide a seamless installation and management experience for NuGet packages in the .NET ecosystem. The use of .NET tools as a deployment mechanism has been a recurring request from users and internal partners who need to authenticate with private repositories. Currently, this solution works for the Windows .NET Framework. However, the goal is to extend this support cross-platform for all supported .NET runtimes.
+
+By implementing this specification, we offer plugin authors the option to use .NET tools for plugin deployment. They can install these as a `tool path` global tool. This eliminates the need to maintain separate versions for `.NET Framework` and `.NET Core`. It also simplifies the installation process by removing the necessity for plugin authors to create subcommands like `codeartifact-creds install/uninstall`.
+
+The ideal workflow for repositories that access private NuGet feeds, such as Azure DevOps, would be:
+
+1. Ensure that the dotnet CLI tools are installed.
+2. Execute the command `dotnet tool install Microsoft.CredentialProviders --tool-path "%UserProfile%/.nuget/plugins/tools"` on the Windows platform. For Linux and Mac, run `dotnet tool install Microsoft.CredentialProviders --tool-path $HOME/.nuget/plugins`.
```

Comment:
Unless the customer is using PowerShell, in which case they need to use a different syntax for the directory. Or on Windows if they're using Git Bash, in which case they use the Linux syntax, although I think there's a good chance that anyone using Git Bash on Windows probably knows enough to figure this out for themselves.

On Linux and Mac, I think there's a good chance that (almost?) every shell uses the `$` syntax for environment variable expansion. Although technically anyone using `pwsh` on Linux/Mac need to use different syntax (again, people using pwsh on Linux/Mac probably know they need to use different syntax).

The point being that needing to use `--tool-path` adds documentation complexity because inexperienced developers who don't know that different shells have different syntax, might run commands that fail, which could lead to bug reports and therefore support workload.

It now occurs to me that I'm being really pedantic about this spec. Sorry ☹️

Reply:
I agree with @zivkan about the awkwardness here. One thing I'd suggest to solve the problem if we do go forward with tools: what if nuget did have a new command `dotnet new credential-provider install ....` that as its implementation just called `dotnet tool install` with the appropriate parameters? That would hide the decision of _where_ the tools needed to be installed from end users.

We already do the inverse in the SDK today, where commands like `dotnet list package` only serve as wrappers around invocations of the core NuGet CLI experience.